### PR TITLE
feat: add affinity-proteomics skill — Olink NPX + SomaLogic SomaScan

### LINKED ISSUE #95: [Skill] Affinity Proteomics Pipeline — Olink NPX + SomaLogic SomaScan
## Skill name

`affinity-proteomics`

## What it does

Provides an end-to-end analysis pipeline for affinity-based proteomics platforms — **Olink** (Proximity Extension Assay, NPX values) and **SomaLogic SomaScan** (SOMAmer aptamers, RFU values). The skill ingests platform-native file formats (Olink NPX CSV/Parquet, SomaLogic ADAT), applies platform-specific QC and normalisation, runs differential abundance analysis with multiple testing correction, and produces volcano plots, heatmaps, pathway enrichment results, and structured reports. It handles the key methodological differences between platforms (log2 NPX vs log10 RFU, bridge normalisation vs hybridisation control normalisation) under a unified interface while preserving platform-specific best practices.

## Bioinformatics tools it wraps

**Olink arm:**
- **OlinkAnalyze** methodology (Olink Proteomics) — NPX QC (IPC checks, sample median, CV%), bridge/reference median normalisation, LOD integration, t-test/ANOVA/linear mixed model differential abundance
- Olink NPX CSV/Parquet readers

**SomaLogic arm:**
- **somadata** Python package (SomaLogic/Canopy) — ADAT file parsing, RFU extraction, clinical metadata handling
- **SomaDataIO** preprocessing methodology — hybridisation control normalisation, median signal normalisation, calibrator QC, MAD-based outlier filtering (≥5% of features exceeding ±6 MAD), log10 transformation
- ColCheck/RowCheck QC flag evaluation

**Shared downstream:**
- **scipy.stats** — t-test, Mann-Whitney U, ANOVA for differential abundance
- **statsmodels** — multiple testing correction (Benjamini-Hochberg FDR, Bonferroni)
- **matplotlib/seaborn** — volcano plots, heatmaps, PCA plots, protein-level QC plots
- **gseapy** (optional) — GO/KEGG/Reactome pathway enrichment on significant proteins

## Input / Output

- **Input**:
  - **Olink**: NPX file (CSV or Parquet from Olink Explore / Explore HT / Target 96) with columns including SampleID, OlinkID, UniProt, Assay, NPX, QC_Warning, LOD; plus a sample metadata CSV (SampleID, Group, Covariates)
  - **SomaLogic**: ADAT file (`.adat` from SomaScan v4.0/v4.1, ~7000 proteins) containing RFU values and embedded clinical metadata; plus optional sample metadata CSV
  - `--platform`: `olink` or `somascan` (auto-detected from file extension if not specified)
  - `--group-col`: Column name for group comparison (e.g. `Disease`, `Treatment`)
  - `--contrast`: Two-group contrast (e.g. `Case,Control`)
  - `--covariates` (optional): Comma-separated covariate columns for adjusted models

- **Output**:
  - **QC report** (`qc/qc_report.md`) — sample-level QC summary (flagged samples, normalisation scale factors, PCA of raw vs normalised data), protein-level QC (CV%, detectability, LOD summary for Olink)
  - **Differential abundance table** (`tables/diff_abundance.tsv`) — per-protein: UniProt ID, gene symbol, log2FC (or NPX/RFU difference), p-value, adjusted p-value (BH-FDR), mean per group, % above LOD (Olink)
  - **Volcano plot** (`figures/volcano.png`) — log2FC vs -log10(adjusted P), with significance thresholds and top protein labels
  - **Heatmap** (`figures/heatmap.png`) — top N differentially abundant proteins with hierarchical clustering
  - **PCA plot** (`figures/pca.png`) — first two principal components coloured by group
  - **Pathway enrichment** (optional, `tables/pathway_enrichment.tsv`) — GO/KEGG/Reactome terms enriched in significant proteins
  - **Structured results** (`result.json`) — summary counts, platform metadata, top hits
  - **Reproducibility bundle** (`reproducibility/`) — exact commands, normalisation parameters, software versions

## Why it matters

Olink and SomaLogic are the two dominant high-throughput affinity proteomics platforms used in large-scale epidemiological and clinical studies (UK Biobank, INTERVAL, deCODE). Together they measure 5,000–11,000 proteins from a single blood sample, making them central to precision medicine and drug target discovery. However, each platform has its own proprietary file format, normalisation method, and QC conventions — requiring researchers to write bespoke analysis scripts for each platform.

The existing ClawBio `proteomics-de` skill handles **mass-spectrometry** LFQ data (MaxQuant/DIA-NN) and explicitly does not cover affinity-based platforms. This new skill fills that gap by providing a unified interface for both Olink and SomaLogic with platform-aware QC, normalisation, and analysis under a single command.

Recent large-scale studies integrating Olink and SomaLogic data ([Eldjarn et al., 2023](https://pubmed.ncbi.nlm.nih.gov/37468513/); [Sun et al., 2023](https://pubmed.ncbi.nlm.nih.gov/37794186/)) demonstrate the importance of standardised, reproducible analysis for both platforms — exactly what this skill provides.

**Who needs this?**
- Epidemiologists analysing UK Biobank Olink Explore or pharma proteomics data
- Clinical researchers running case-control or longitudinal studies with SomaScan panels
- Drug target discovery teams integrating pQTL results across platforms
- Core facilities standardising affinity proteomics analysis workflows

### Recommended QC thresholds (configurable via CLI)

**Olink:**

| Parameter | Default | Source |
|-----------|---------|--------|
| Sample QC warning | Exclude samples with QC_Warning = "WARN" | Olink best practice |
| IPC deviation | > 0.3 NPX from plate median → flag | OlinkAnalyze |
| Protein detectability | Report % samples above LOD per protein | Olink Explore guidance |
| Bridge normalisation | Median paired difference adjustment | OlinkAnalyze `olink_normalization` |

**SomaLogic:**

| Parameter | Default | Source |
|-----------|---------|--------|
| Normalisation acceptance | 0.4 ≤ scale factor ≤ 2.5 | SomaDataIO `preProcessAdat` |
| Outlier filter | ≥5% features at ±6 MAD + 5x fold-change from median | SomaDataIO (plasma/serum) |
| Transformation | log10(RFU) | SomaLogic best practice |
| ColCheck/RowCheck | Exclude flagged analytes/samples | SomaScan QC protocol |

**Shared:**

| Parameter | Default | Source |
|-----------|---------|--------|
| FDR threshold | Adjusted P < 0.05 (Benjamini-Hochberg) | Standard practice |
| Fold-change threshold | |log2FC| > 0.25 (configurable) | Common convention |
| Top N for heatmap | 50 proteins | Default, configurable |
| Pathway databases | GO Biological Process + KEGG | gseapy defaults |

### Key references

- [Assarsson et al. (2014)](https://pubmed.ncbi.nlm.nih.gov/25057488/) — Olink PEA technology. *PLOS ONE* 9:e95192
- [Gold et al. (2010)](https://pubmed.ncbi.nlm.nih.gov/20829826/) — SOMAmer aptamer technology. *PLOS ONE* 5:e15004
- [Eldjarn et al. (2023)](https://pubmed.ncbi.nlm.nih.gov/37468513/) — Large-scale plasma proteomics with Olink Explore 3072. *Nature* 622:727
- [Sun et al. (2023)](https://pubmed.ncbi.nlm.nih.gov/37794186/) — Plasma proteome atlas across Olink and SomaScan. *Nature* 622:340
- [Pietzner et al. (2021)](https://pubmed.ncbi.nlm.nih.gov/33067605/) — Mapping the proteo-genomic convergence of human diseases. *Science* 374:eabj1541
- [OlinkAnalyze R package](https://cran.r-project.org/package=OlinkAnalyze) — Official Olink analysis toolkit
- [SomaDataIO R package](https://somalogic.github.io/SomaDataIO/) — Official SomaLogic data I/O
- [somadata Python package](https://pypi.org/project/somadata/) — Python ADAT parser (SomaLogic/Canopy)

## Are you willing to implement it?

- [x] Yes, I'd like to build this skill
- [ ] No, I'm proposing it for someone else to pick up

### PR DESCRIPTION
**Pull request by RezaJF**
## Summary

- Add **affinity-proteomics** skill for unified analysis of affinity-based proteomics platforms (Olink PEA and SomaLogic SomaScan)
- Dual-platform pipeline: platform-specific parsing (NPX CSV / ADAT), QC (QC_Warning / RowCheck+ColCheck), normalisation, and shared downstream analysis (differential abundance with BH-FDR, volcano plot, heatmap, PCA)
- Fills the gap between the existing `proteomics-de` skill (mass-spec LFQ) and the two dominant affinity platforms used in UK Biobank, INTERVAL, and deCODE studies

## Skill affected

- New skill: `affinity-proteomics` (CLI alias: `affprot`)

## Checklist

- [x] SKILL.md is present and complete (YAML frontmatter + methodology)
- [x] Tests included and passing (`pytest -v`) — 19 tests (unit + integration)
- [x] Demo data included — synthetic Olink NPX (80 samples, 40 proteins) + SomaLogic uses somadata's bundled ADAT
- [x] No patient/sensitive data committed
- [x] Follows CONTRIBUTING.md conventions
- [x] Registered in `clawbio.py` SKILLS dict
- [x] Test path added to `pytest.ini`
- [x] `skills/catalog.json` regenerated (42 skills)

## Test output

```
$ python -m pytest skills/affinity-proteomics/tests/ -v
19 passed in 6.55s
```

Test classes:
- `TestOlinkParser` (6): NPX parsing, sample/protein counts, QC summary, metadata
- `TestSomaLogicParser` (4): ADAT parsing, sample filtering, log10 transform, QC
- `TestDifferentialAbundance` (5): result structure, p-value sorting, DE protein recovery, effect direction, error handling
- `TestOlinkDemo` (2): end-to-end demo, report content validation
- `TestSomaLogicDemo` (2): end-to-end SomaScan demo, KLK3/PSA detection in F vs M

## Demo output

**Olink** (synthetic data, Case vs Control):
```
[PROT] Samples after QC: 78, Proteins: 40
[PROT] Significant proteins (FDR<0.05, |log2FC|>=0.25): 6
  Top hit: IL1B (P=5.75e-12), recovers all 5 ground-truth DE proteins
```

**SomaLogic** (bundled SomaScan example, F vs M):
```
[PROT] Samples after QC: 168, Proteins: 5045
[PROT] Significant proteins (FDR<0.05, |log2FC|>=0.25): 15
  Top hit: KLK3/PSA (P=2.98e-42) — known sex-differential protein
```

## Dependencies

- `somadata` >= 1.2 (SomaLogic ADAT parsing)
- `scipy`, `statsmodels` (statistical testing)
- `matplotlib`, `seaborn` (visualisation)

## Key references

- Assarsson et al. (2014) PLOS ONE 9:e95192 (Olink PEA)
- Gold et al. (2010) PLOS ONE 5:e15004 (SOMAmer)
- Eldjarn et al. (2023) Nature 622:727 (Olink Explore at scale)
- Sun et al. (2023) Nature 622:340 (cross-platform proteome atlas)

Closes #95

### CODE REVIEW

**Review by manuelcorpas — APPROVED** [review]
## ✅ ClawBio PR Audit: SAFE

| Category | Count |
|----------|-------|
| Critical | 0 |
| Warning | 2 |
| Info | 3 |

Well-structured affinity-proteomics skill (Olink NPX + SomaLogic SomaScan). No dangerous calls, no network access, comprehensive tests. Contributor is a statistical geneticist at 23andMe.

Minor notes: catalog.json description field shows `>-` instead of resolved text (catalog generation bug), and a stowaway fine-mapping entry appeared in catalog.json.

---
*Automated audit by ClawBio pr-audit. Manual review is still recommended.*

**Review by manuelcorpas — CHANGES_REQUESTED** [review]
## PR Audit — CHANGES REQUESTED 🔧

**0 critical, 3 warnings (1 is a real bug), 6 info.** No security issues. One statistical bug needs fixing.

### Bug: SomaLogic log2FC calculation (W2)
File: `affinity_proteomics.py`, line 448.

The current code:
```python
log2fc = (mean1 - mean2) / math.log(2) if data.platform == "somascan" else mean1 - mean2
```

For SomaLogic data (log10 scale), the conversion to log2 fold change should divide by `log10(2) = 0.3010`, not `ln(2) = 0.6931`. This **overstates fold changes by approximately 2.3x** for all SomaLogic differential expression results.

Fix: change `math.log(2)` to `math.log10(2)`.

The Olink branch (`mean1 - mean2`) is correct since NPX is already log2 scale.

### Other warnings (non-blocking)
1. **Malformed catalog descriptions** — `">-"` YAML leak in catalog.json entries.
2. **sklearn unlisted dependency** — PCA function imports `sklearn.decomposition.PCA` but scikit-learn is not in SKILL.md dependencies. Function gracefully fails on ImportError, but PCA plots silently won't generate.

### What looks good
- No network calls, no eval/exec, no subprocess, no hidden unicode, no secrets
- CSV demo data is clean synthetic Olink NPX (80 samples, 40 proteins, real UniProt IDs)
- All computation local-only, disclaimer in all reports
- 214 lines of tests
- References Assarsson et al. (2014) and Gold et al. (2010)

Automated audit by ClawBio pr-audit. Manual review is still recommended.

**Review by manuelcorpas — APPROVED** [review]
Re-audit after the 2026-04-05 push: the critical SomaLogic log2FC bug is **fixed** — `math.log10(2)` is now used at `skills/affinity-proteomics/affinity_proteomics.py:452`. Nice catch-and-fix.

Outstanding minor items (non-blocking, can be follow-ups):
- `sklearn.decomposition.PCA` is still imported without scikit-learn being listed as a SKILL.md dependency. Function degrades gracefully on ImportError so PCA plots silently skip, but listing it as optional in the SKILL.md `dependencies` section would be clearer for users.
- SKILL.md still missing `## Workflow` and `## Output Structure` sections per the skill template.

Security audit: SAFE. 0 critical, 0 warnings, 2 info. 9 files changed.

Approving. Moving to merge.

Automated audit by ClawBio pr-audit. Manual review is still recommended.

**Comment by manuelcorpas** [comment]
Merge conflicts against main — I approved the re-audit but cannot merge. Please rebase onto the current main and I'll ship it. 🙏

**Comment by RezaJF** [comment]
Hi @manuelcorpas- Rebased onto current main (`3adb884`). Merge conflict resolved.

**Re: CI test failures** — the 2 failing tests (`test_dpyd_normal`, `test_tpmt_normal`) are in `skills/pharmgx-reporter/`, which this PR does not touch. They are a pre-existing issue from the CPIC compliance fix (`f74be38`) where the phenotype logic now requires 3/3 SNPs for DPYD and TPMT but the demo test data only provides 2/3 SNPs. This PR's own 19 tests all pass.

**Comment by manuelcorpas** [comment]
This PR has merge conflicts with main (likely from shared `catalog.json`, `clawbio.py`, or `lit-synthesizer/SKILL.md` changes). Could you rebase onto main? The audit review is approved — will merge once conflicts are resolved.

**Comment by manuelcorpas** [comment]
Same conflict as #92 — catalog.json overlap from #129 merge. One more rebase needed. Approved, will merge immediately.

**Comment by manuelcorpas** [comment]
@RezaJF This PR is approved and ready to merge, but it now has merge conflicts from recent upstream changes. Could you please rebase onto current main and push?

```bash
git fetch upstream
git rebase upstream/main
git push --force-with-lease
```

Will merge as soon as conflicts are resolved.
